How-to-upgrade-review-version.md

```diff
@@ -0,0 +1,399 @@
+    type Options = {
+        topLevel: boolean;
+        rest: boolean;
+        assignments: any[];
+    };
+    class PatternVisitor {
+        constructor(
+            options: any,
+            rootPattern: any,
+            callback: (p: any, options: Options) => void);
+        rightHandNodes: Array<Expression | SpreadElement>;
+
+        Identifier(node: Node): void;
+        visit(node: Node): void;
+    }
+    export = PatternVisitor;
+}
+```
+
+I got this by (1) looking at pattern-visitor.js and (2) the estree
+types. eslint-scope uses estree types without explicitly referring to
+them, so an eventual correct fix would change eslint-scope. For now, I
+have enough to keep going.
+
+## Fix errors in existing types
+
+Unfortunately, the improved type for pattern-visitor once again causes
+an error on `new PatternVisitor`. This time, the callback's type,
+`Function` isn't specific enough to work with the specific function
+type of the callback:
+
+``ts
+callback: (p: any, options: Options) => void
+```
+
+So the existing JSDoc type annotation needs to change:
+
+```js
+    /**
+     * Override to use PatternVisitor we overrode.
+     * @param {Identifier} node The Identifier node to visit.
+     * @param {Object} [options] The flag to visit right-hand side nodes.
+     * @param {Function} callback The callback function for left-hand side nodes.
+     * @returns {void}
+     */
+    visitPattern(node, options, callback) {
+```
+
+I changed `@param {Function} callback` to the more precise
+`@param {(p: any, options: Options) => void} callback`.
+
+Except that Options isn't exported by my pattern-visitor workaround.
+So I moved it in the global namespace, outside the `declare module`:
+
+```ts
+type Options = {
+    topLevel: boolean;
+    rest: boolean;
+    assignments: any[];
+};
+
+declare module "eslint-scope/lib/pattern-visitor" {
+    import { Node, Expression, SpreadElement } from "estree";
+    class PatternVisitor {
+        constructor(
+            options: any,
+            rootPattern: any,
+            callback: (p: any, options: Options) => void);
+        rightHandNodes: Array<Expression | SpreadElement>;
+
+        Identifier(node: Node): void;
+        visit(node: Node): void;
+    }
+    export = PatternVisitor;
+}
+```
+
+This is messy but at the same time *horribly* convenient.
+
+## Add type annotations to everything else
+
+Once you get all the existing type annotations working, the next step
+is to add type annotations to everything else. You can turn on
+"strict": true to see how far you have to go. You should fall into a
+back-and-forth of adding new type annotations and fixing old ones, or
+fixing bugs.
+
+This *usually* means adding type annotations and types to support
+them. You won't have to change the actual code much unless you find a
+bug. If you want to see how far you have to go, turn on "strict" and
+see how many errors you get.
+
+You can use the infer-from-usage suggestions to help.
```

Comment:
Do you have a code bug you found with this method? It would be nice to call out an example, otherwise it's not clear what would motivate me to do all this work

Reply:
No :( I'll see if I can find one.